src/components/browser/requestOperations/Swap.tsx

```diff
@@ -0,0 +1,214 @@
+import {ActiveAccount, KeyTypes} from 'actions/interfaces';
+import {showModal} from 'actions/message';
+import {RequestSwap} from 'hive-keychain-commons';
+import React, {useEffect, useState} from 'react';
+import {MessageModalType} from 'src/enums/messageModal.enums';
+import {TransactionOptions} from 'src/interfaces/multisig.interface';
+import {getCurrency} from 'utils/hive';
+import {getAccount} from 'utils/hiveUtils';
+import {RequestId} from 'utils/keychain.types';
+import {translate} from 'utils/localize';
+import {SwapTokenUtils} from 'utils/swap-token.utils';
+import {getTokenPrecision} from 'utils/tokens.utils';
+import RequestItem from './components/RequestItem';
+import RequestOperation from './components/RequestOperation';
+import {RequestComponentCommonProps} from './requestOperations.types';
+
+type Props = {
+  request: RequestSwap & RequestId;
+} & RequestComponentCommonProps;
+
+export default ({
+  request,
+  accounts,
+  closeGracefully,
+  sendResponse,
+  sendError,
+}: Props) => {
+  const {request_id, ...data} = request;
+  const {username, startToken, endToken, amount, steps, slippage} = data;
+  const [estimateValue, setEstimateValue] = useState<number>();
+  const [balance, setBalance] = useState<number>();
+  const [balanceAfterSwap, setBalanceAfterSwap] = useState<number>();
+
+  useEffect(() => {
+    const calculateEstimate = async () => {
+      try {
+        const swapConfig = await SwapTokenUtils.getConfig();
+        if (!swapConfig) return;
+
+        const lastStep = steps[steps.length - 1];
+        if (!lastStep) return;
+
+        const precision = await getTokenPrecision(lastStep.endToken);
+        const value = Number(lastStep.estimate);
+        const fee = (value * swapConfig.fee.amount) / 100;
+        const finalValue = Number(value - fee).toFixed(precision);
+        setEstimateValue(Number(finalValue));
+      } catch (error) {
+        console.error('Error calculating estimate:', error);
+      }
+    };
+
+    calculateEstimate();
+  }, [steps]);
+
+  const getBalance = async () => {
+    const account = accounts.find((e) => e.name === username);
+    if (!account) return 0;
+
+    try {
+      const extendedAccount = await getAccount(username);
+      if (!extendedAccount) return 0;
+
+      let balance = 0;
+      if (startToken === getCurrency('HBD')) {
+        balance = parseFloat(
+          (extendedAccount.hbd_balance as string).split(' ')[0],
+        );
+      } else if (startToken === getCurrency('HIVE')) {
+        balance = parseFloat((extendedAccount.balance as string).split(' ')[0]);
+      } else {
+        const tokenBalances = await SwapTokenUtils.getSwapTokenStartList(
+          extendedAccount,
+        );
+        if (tokenBalances && tokenBalances.length > 0) {
+          const tokenBalance = tokenBalances.find(
+            (token) => token.symbol === startToken,
+          );
+          if (tokenBalance) {
+            balance = parseFloat(tokenBalance.balance);
```

Comment:
balance (and estimate) must be trimmed to:
- HIVE / HBD : 3 decimals
- Hive Engine token : `precision` field 

Remove trailing zero decimals.

We most likely already have a formatting util for this